lib/search_files.js

```diff
@@ -0,0 +1,252 @@
+    if (ret) {
+      return ret;
+    }
+  }
+  return null;
+}
+
+function eachUnNullElem(arr, fn /*, ...*/) {
+  var args = Array.prototype.slice.call(arguments, 2);
+
+  var ret = [];
+  for (var i = 0, n = arr.length; i < n; i++) {
+    var v = fn.apply(arr, [arr[i]].concat(args));
+    if (v) {
+      if (isArray(v)) {
+        Array.prototype.push.apply(ret, v);
+      } else {
+        ret.push(v);
+      }
+    }
+  }
+  return ret;
+}
+
+
+function normalizeHomedir(pth, homedir) {
+  if (!pth) {
+    return pth;
+  }
+
+  if (pth[0] === '~') {
+    if (pth.length === 1) {
+      return homedir;
+    }
+
+    if (pth[1] === '/') {
+      return path.join(homedir, pth.slice(2));
+    }
+
+    if (isWindows && pth[1] === '\\') {
+      return path.join(homedir, pth.slice(2));
+    }
+  }
+
+  return pth;
+}
+
+function normalizePath(pth, ctx) {
+  if (isFunction(pth)) {
+    pth = pth();
+  }
+
+  pth = normalizeHomedir(pth, ctx.homedir);
+
+  if (!isAbsolutePath(pth)) {
+    return path.join(ctx.cwd, pth);
+  }
+
+  return path.normalize(pth);
+}
+
+
+var initCtx = {
+  homedir: oshomedir(),
+  cwd: process.cwd(),
+};
+
+function normalizeContext(ctx) {
+  ctx = mapValues(ctx || {}, doNothing);
+
+  ctx.cwd = !ctx.cwd ? initCtx.cwd :
+    normalizePath(ctx.cwd, initCtx);
+
+  ctx.homedir = !ctx.homedir ? initCtx.homedir :
+    normalizePath(ctx.homedir, initCtx);
+
+  return ctx;
+}
+
+function normalizeExtensions(exts) {
+  if (isFunction(exts)) {
+    exts = exts();
+  }
+
+  if (isString(exts)) {
+    return [exts];
+  }
+
+  if (isArray(exts)) {
+    return exts.filter(isString);
+  }
+
+  if (isPlainObject(exts)) {
+    return Object.keys(exts).sort();
+  }
+
+  return [''];
+}
+
+function findFile(pth, basedir, findUp, opeArr) {
+
+  var parsed = parsePath(pth);
+  if (parsed.isAbsolute) {
+    findUp = false;
+    basedir = parsed.root;
+    pth = pth.slice(parsed.root.length);
+  } else if (parsed.root) {  // pth has a drive letter on windows.
+    basedir = path.resolve(parsed.root);
+    pth = pth.slice(parsed.root.length);
+  }
+
+  if (!findUp) {
+    pth = path.join(basedir, pth);
+    if (fs.existsSync(pth)) {
+      return opeArr([pth], checkSymbolicLink);
+    }
+    return opeArr([]);
+  }
+
+  var lastDir;
+  do {
+    var fp = path.join(basedir, pth);
+    if (fs.existsSync(fp)) {
+      return opeArr([fp], checkSymbolicLink);
+    }
+
+    lastDir = basedir;
+    basedir = path.dirname(basedir);
+  } while (lastDir !== basedir);
+
+  return opeArr([]);
+}
+
+function checkSymbolicLink(pth) {
+  if (!fs.lstatSync(pth).isSymbolicLink()) {
+    return pth;
+  }
+
+  try {
+    if (fs.existsSync(fs.realpathSync(pth))) {
+      return pth;
+    }
+  } catch (e) {}
+
+  return null;
+}
+
+
+function search(opts, ctx) {
+  ctx = normalizeContext(ctx);
+  return mapValues(opts || {}, searchWithOpt);
+
+  function searchWithOpt(opt, optname) {
+    if (!opt || isString(opt) || isArray(opt)) {
```

Comment:
I think the `searchPaths` option is not mandatory. 
If `configFiles` option was null, to exit early is also OK because null can be regarded as specifying an empty array. But if an empty string, I think it better to use the only option name as a path.